src/zon.zig

```diff
@@ -0,0 +1,831 @@
+    } });
+
+    const many_item_ptr_type = try ip.get(gpa, self.sema.pt.tid, .{ .ptr_type = .{
+        .child = ptr_info.child,
+        .sentinel = ptr_info.sentinel,
+        .flags = b: {
+            var flags = ptr_info.flags;
+            flags.size = .Many;
+            break :b flags;
+        },
+        .packed_offset = ptr_info.packed_offset,
+    } });
+
+    const many_item_ptr = try ip.get(gpa, self.sema.pt.tid, .{
+        .ptr = .{
+            .ty = many_item_ptr_type,
+            .base_addr = .{
+                .uav = .{
+                    .orig_ty = res_ty.toIntern(),
+                    .val = array,
+                },
+            },
+            .byte_offset = 0,
+        },
+    });
+
+    const len = (try self.sema.pt.intValue(Type.usize, elems.len)).toIntern();
+
+    return ip.get(gpa, self.sema.pt.tid, .{ .slice = .{
+        .ty = res_ty.toIntern(),
+        .ptr = many_item_ptr,
+        .len = len,
+    } });
+}
+
+fn lowerUnion(self: LowerZon, node: Zoir.Node.Index, res_ty: Type) !InternPool.Index {
+    const ip = &self.sema.pt.zcu.intern_pool;
+    try res_ty.resolveFields(self.sema.pt);
+    const union_info = self.sema.pt.zcu.typeToUnion(res_ty).?;
+    const enum_tag_info = union_info.loadTagType(ip);
+
+    const field_name, const maybe_field_node = switch (node.get(self.file.zoir.?)) {
+        .enum_literal => |name| b: {
+            const field_name = try ip.getOrPutString(
+                self.sema.gpa,
+                self.sema.pt.tid,
+                name.get(self.file.zoir.?),
+                .no_embedded_nulls,
+            );
+            break :b .{ field_name, null };
+        },
+        .struct_literal => b: {
+            const fields: std.meta.fieldInfo(Zoir.Node, .struct_literal).type = switch (node.get(self.file.zoir.?)) {
+                .struct_literal => |fields| fields,
+                else => return self.fail(
+                    .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                    "expected type '{}'",
+                    .{res_ty.fmt(self.sema.pt)},
+                ),
+            };
+            if (fields.names.len != 1) {
+                return self.fail(
+                    .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+                    "expected type '{}'",
+                    .{res_ty.fmt(self.sema.pt)},
+                );
+            }
+            const field_name = try ip.getOrPutString(
+                self.sema.gpa,
+                self.sema.pt.tid,
+                fields.names[0].get(self.file.zoir.?),
+                .no_embedded_nulls,
+            );
+            break :b .{ field_name, fields.vals.at(0) };
+        },
+        else => return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "expected type '{}'",
+            .{res_ty.fmt(self.sema.pt)},
+        ),
+    };
+
+    const name_index = enum_tag_info.nameIndex(ip, field_name) orelse {
+        return self.fail(
```

Comment:
I just thought; why does `fail` not just take the ZOIR node, and do `getAstNode` there? That's the only place you ever get the node from, after all.

Reply:
Turns out there are a few places where I need to get the AST node sooner because I need to get the location of e.g. a strut field. I made this change anyway though since it makes much of the code nicer, and the cases where something different happens are now easy to find with ctrl+f.